### PR TITLE
OpsGenie

### PR DESCRIPTION
**Pull request by bziser**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://github.com/demisto/etc/issues/20166

## Description
To enhance the current OpsGenie integration to include alert and incident management functions as defined in OpsGenie's API documentation.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [X] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/OpsGenie/Classifiers/classifier-OpsGenie_-_Classifier.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+	"brands": null,
+	"defaultIncidentType": "Unclassified",
+	"definitionId": "",
+	"description": "",
+	"feed": false,
+	"fromServerVersion": "",
+	"id": "e0618001-7d8e-4832-8c87-a99f72ce440d",
```
Comment (bakatzir) [inline Packs/OpsGenie/Classifiers/classifier-OpsGenie_-_Classifier.json:8]: id should be == name

File: Packs/OpsGenie/Classifiers/classifier-OpsGenie_-_Mapper.json [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    "complex": null,
+                    "simple": "createdAt"
+                }
+            }
+        }
+    },
+    "name": "OpsGenie - Mapper",
+    "type": "mapping-incoming",
+    "version": -1,
+    "fromVersion": "6.0.0"
```
Comment (bakatzir) [inline Packs/OpsGenie/Classifiers/classifier-OpsGenie_-_Mapper.json:166]: as the integration is fromversion: '6.2.0', please have this one as well.

File: Packs/OpsGenie/IncidentTypes/incidenttype-OpsGenie_Alert.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    "autorun": false,
+    "disabled": false,
+    "reputationCalc": 0,
+    "onChangeRepAlg": 0,
+    "detached": false,
+    "extractSettings": {
+        "mode": "Specific",
+        "fieldCliNameToExtractSettings": {}
+    },
+    "fromVersion": "5.5.0"
```
Comment (bakatzir) [inline Packs/OpsGenie/IncidentTypes/incidenttype-OpsGenie_Alert.json:27]: as the integration is fromversion: '6.2.0', please have this one as well.

File: Packs/OpsGenie/IncidentTypes/incidenttype-OpsGenie_Incident.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "autorun": false,
+    "disabled": false,
+    "reputationCalc": 0,
+    "onChangeRepAlg": 0,
+    "detached": false,
+    "extractSettings": {
+        "mode": "Specific",
+        "fieldCliNameToExtractSettings": {}
+    },
+    "fromVersion": "5.5.0"
```
Comment (bakatzir) [inline Packs/OpsGenie/IncidentTypes/incidenttype-OpsGenie_Incident.json:27]: as the integration is fromversion: '6.2.0', please have this one as well.

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_description.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+## BaseIntegration Help
+
+Markdown file for integration configuration  help snippet. In this file add:
+- Brief information about how to retrieve the API key of your product
+- Other useful information on how to configure your integration in XSOAR
+
+Since this is a Markdown file, we encourage you to use MD formatting for sections, sub-sections, lists, etc.
```
Comment (bakatzir) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_description.md:7]: TODO

File: Packs/OpsGenie/ReleaseNotes/2_0_0.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (bakatzir) [inline Packs/OpsGenie/ReleaseNotes/2_0_0.md:1]: fix the available from

File: Packs/OpsGenie/README.md [hunk 7]
```diff
@@ -0,0 +1 @@
+X
```
Comment (bakatzir) [inline Packs/OpsGenie/README.md:1]: TODO

File: Packs/OpsGenie/pack_metadata.json [hunk 8]
```diff
@@ -2,7 +2,7 @@
   "name": "OpsGenie",
   "description": "Get current on-call assignments, schedules, and users info",
```
Comment (bakatzir) [inline Packs/OpsGenie/pack_metadata.json:3]: let's re work that with the docs team

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml [hunk 9]
```diff
@@ -0,0 +1,1612 @@
...
+      type: String
+  dockerimage: demisto/python3:3.9.7.24076
+  isfetch: true
+  runonce: false
+  script: '-'
+  type: python
+  subtype: python3
+fromversion: '6.2.0'
+tests:
+- No tests (auto formatted)
```
Comment (bakatzir) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml:1612]: you do have one
Packs/OpsGenie/TestPlaybooks/playbook-OpsGenieV3-Test.yml

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml [hunk 10]
```diff
@@ -0,0 +1,1612 @@
+category: Case Management
+commonfields:
+  id: OpsGenieV3
```
Comment (bakatzir) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml:3]: why not have it same as the displayname `OpsGenie v3`?
just for same string all around.

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py [hunk 11]
```diff
@@ -0,0 +1,1015 @@
+import traceback
+import dateparser
+from datetime import datetime
+from typing import Dict, Any, Callable, Tuple
```
Comment (bakatzir) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py:4]: we have these already in Commonserverpython

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py [hunk 12]
```diff
@@ -0,0 +1,1015 @@
...
+import demistomock as demisto
+from CommonServerPython import *  # noqa # pylint: disable=unused-wildcard-import
+from CommonServerUserPython import *  # noqa
+
+import requests
```
Comment (bakatzir) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py:9]: same

**Review by bakatzir — COMMENTED** [review]
please:
1. deprecate Packs/OpsGenie/Integrations/integration-OpsGenie.yml
2. do Packs/OpsGenie/README.md with docs team and pack desc

I Have not reviewed the py files @yuvalbenshalom .

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py (hunk 11 above)
Comment (bziser) [reply Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py:4]: from typing import Callable, Tuple
These aren't in Commonserverpython

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml (hunk 10 above)
Comment (bziser) [reply Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml:3]: If I change the is I need to change also the name of the yml file. Is it necessary?

Comment (bakatzir) [reply Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.yml:3]: no, just easier to have id==name==displayname

**Comment by ShirleyDenkberg** [comment]
@bziser 
- The content of the Packs/OpsGenie/README.md file is not correct.  This is the pack README. It is supposed to only contain a brief summary of what the pack does. It appears in Marketplace when you click the Opsgenie tile.  It is not the integration README.
- Please regenerate the integration README for OpsGenie v3 since I made several changes to the yml file.  However, I edited the Breaking changes section in the README, so do not overwrite that.

@bakatzir @yuvalbenshalom Doc review completed.

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+    mock_client = OpsGenieV3.Client(base_url="")
+    mock_client.get_team = mock.MagicMock()
+    OpsGenieV3.get_teams(mock_client, {"team_id": 1234})
+    assert mock_client.get_team.called
+    mock_client.list_teams = mock.MagicMock()
+    OpsGenieV3.get_teams(mock_client, {})
+    assert mock_client.list_teams.called
+
+
+def test_fetch_incidents_command(mocker):
```
Comment (yuvalbenshalom) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py:354]: please add a test with no results. check that last run is calculated correctly

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+    mocker.patch('CommonServerPython.get_demisto_version', return_value={"version": "6.2.0"})
+    mock_client = OpsGenieV3.Client(base_url="")
+    mocker.patch.object(mock_client, 'list_alerts',
+                        return_value=util_load_json('test_data/get_alerts.json'))
+    mocker.patch.object(mock_client, 'list_incidents',
+                        return_value=util_load_json('test_data/get_incidents.json'))
+    res, last_run = OpsGenieV3.fetch_incidents_command(mock_client, {"max_fetch": 1})
+    assert len(res) == 2
```
Comment (yuvalbenshalom) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py:362]: please check that last run is calculated as you expected.

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+
+
+def test_fetch_with_paging(mocker):
```
Comment (yuvalbenshalom) [inline Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3_test.py:365]: please add a similar case with incidents

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
looks good.
See my comments.

File: Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py [hunk 16]
```diff
@@ -0,0 +1,1067 @@
...
+
+
+def fetch_incidents_by_type(client: Client,
+                            params: Dict[str, Any],
+                            incident_fetching_func: Callable,
+                            now: datetime,
+                            last_run_dict: Optional[dict] = None):
+    query = params.get('query')
+    limit = int(params.get('max_fetch', 50))
+    fetch_time = params.get('first_fetch', '3 days').strip()
```
[the remark this answers is not in the document]
Comment (bziser) [reply Packs/OpsGenie/Integrations/OpsGenieV3/OpsGenieV3.py:916]: But I'm passing next the params to the incident_fetching_func

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**